src/consensus/tx_verify.cpp

```diff
@@ -22,11 +22,16 @@
 
 bool IsFinalTx(const CTransactionRef &tx, int nBlockHeight, int64_t nBlockTime)
 {
-    if (tx->nLockTime == 0)
+    return IsFinalTx(*tx, nBlockHeight, nBlockTime);
+}
+
+bool IsFinalTx(const CTransaction &tx, int nBlockHeight, int64_t nBlockTime)
+{
```

Comment:
we use CTransactionRef's everywhere now, why the need for the change to using CTransaction?  Same with CheckFinalTx() (they are never called anywhere using just a CTransaction)

Reply:
converting the CTransaction* into a CTransactionRef was killing performance

Reply:
Still, the function is never called anywhere using just a CTransaction(), remember we moved everything over to Ref's a while back?  Futhermore I'm wondering if you're thinking of make_shared() instead of MakeTransactionRef() which forwards a reference so there's not hit to performance.

Reply:
This is new code, we didn't use refs back then.  I did a bit of optimization work while doing this port and this problem jumped out.  

@ptschip somehow we have a pointer, not a ref in this code, perhaps because the wallet uses a cwallettransaction and casts to its base class.

Reply:
Ok, I'm wondering then if this is really the problem, the definition is incorrect, we're really not supposed to be passing a smart pointer as a reference into a function... and this is a problem I noticed a while ago from code we inherited from Core and then I ended up making the same error myself...I removed it in a few places but didn't do it here.

so the following:
-bool IsFinalTx(const CTransactionRef &tx, int nBlockHeight, int64_t nBlockTime)
should be:
+bool IsFinalTx(const CTransactionRef tx, int nBlockHeight, int64_t nBlockTime)

Do you think this might account for the performance issue you said you were seeing?

Reply:
no the problem was passing through the CTransaction constructor again because making a CTransactionRef from a CTransaction* needs to allocate a CTransaction on the stack and copy the CTransaction* to the memory inside the CTransactionRef.